### PR TITLE
configmap & utime sync: provide via hive cell

### PR DESCRIPTION
**Pull request by mhofstetter**
Currently, the interaction with the BPF map configmap and the corresponding utime sync is mainly via global accessors.

This commit refactors this, by introducing hive cells for the configmap and the utime sync.

### CODE REVIEW

File: pkg/datapath/linux/utime/cell.go
Comment (mhofstetter) [inline pkg/datapath/linux/utime/cell.go:31]: @joamaki would it be worth to have a provider for the controller manager?

Comment (joamaki) [reply pkg/datapath/linux/utime/cell.go:31]: What do you think of https://github.com/cilium/cilium/pull/24558?

File: pkg/maps/configmap/cell.go
Comment (joamaki) [inline pkg/maps/configmap/cell.go]: Forced instantiation really required?

File: pkg/maps/configmap/cell.go
Comment (joamaki) [inline pkg/maps/configmap/cell.go]: Stop hook missing that closes it

File: pkg/datapath/linux/utime/cell.go
Comment (mhofstetter) [reply pkg/datapath/linux/utime/cell.go:31]: interesting - will take a look.

i think something like this could help in this case - and could be refactored in the future (once its merged)

File: pkg/maps/configmap/cell.go
Comment (mhofstetter) [reply pkg/maps/configmap/cell.go]: no not really - having the utime sync depending on it - which triggers the instantiation.

but i thought it would be best to instantiate the map properly so the datapath can rely on it. otherwise it kind of always needs someone who depends on it to get initialized. or does datapath code create the map as well and would work without this initialization from userspace?

File: pkg/maps/configmap/cell.go
Comment (mhofstetter) [reply pkg/maps/configmap/cell.go]: thanks - added!

File: pkg/maps/configmap/cell.go
Comment (jrajahalme) [reply pkg/maps/configmap/cell.go]: It looks to me that we do not create any bpf maps from the datapath?

**Review by jrajahalme — APPROVED** [review]
LGTM

**Comment by mhofstetter** [comment]
rebased and resolved conflicts

File: pkg/maps/configmap/cell.go
Comment (mhofstetter) [reply pkg/maps/configmap/cell.go]: introduced a bpf registry where map cells register themselves by using the value group `bpfMaps`.

this way datapath constructor can depend on the bpf map registry and the initialization order is ensured.

it's just a start which addresses this issue. might be enriched / replaced in the future.

@joamaki can you take another look :eyes:

File: pkg/maps/mapregistry/cell.go
Comment (joamaki) [inline pkg/maps/mapregistry/cell.go]: I would put this file into `pkg/maps/cell.go`. I don't think we need a separate package for it. There it also makes more sense that it references all the maps underneath it.

File: pkg/maps/configmap/cell.go
Comment (joamaki) [inline pkg/maps/configmap/cell.go]: Convention for group struct tags is to use dashed names: `bpf-maps`

File: pkg/maps/mapregistry/cell.go
Comment (joamaki) [inline pkg/maps/mapregistry/cell.go]: I would just add to datapathParams:
```
Maps []maps.BpfMap `group:"bpf-maps"`
```

This wrapper seems like an unnecessary indirection.

File: pkg/maps/configmap/cell.go
Comment (joamaki) [inline pkg/maps/configmap/cell.go:38]: I usually try to avoid writing `cell.Out`s with group values for the fear of typoing the group name. You can get around this with a helper, e.g. `func NewMapOut[Map any](m Map) MapOut[Map]`, but perhaps not worth adding it just yet.

File: pkg/maps/mapregistry/cell.go
Comment (joamaki) [reply pkg/maps/mapregistry/cell.go]: To break the cycle to `BpfMap` move that to `pkg/maps/types` ?
EDIT: or to pkg/bpf. I think eventually that's where we'd have the canonical bpf map type anyway. e.g. I'd expect we'd have `Loader` depending on all the maps and then rewriting references using the open maps rather than pin paths, so this `BpfMap` becomes a real thing rather than a dependency marker.

Comment (mhofstetter) [reply pkg/maps/mapregistry/cell.go]: yeah, the actual reason was the cycle. but makes much more sense by having cells in `pkg/maps` & the type in `pkg/bpf`. thanks

File: pkg/maps/configmap/cell.go
Comment (mhofstetter) [reply pkg/maps/configmap/cell.go]: thanks, fixed

File: pkg/maps/mapregistry/cell.go
Comment (mhofstetter) [reply pkg/maps/mapregistry/cell.go]: had it this way when starting out, but felt like exposing too many internals by depend on the value group name alone. but definitely agree that the indirection otherwise doesn't make sense -> removed

**Review by joamaki — APPROVED** [review]

**Comment by mhofstetter** [comment]
/test-1.27-net-next

File: pkg/maps/cells.go
Comment (lmb) [inline pkg/maps/cells.go:13]: When should I add something here vs. to the datapath cell as done for the configmap?

Comment (lmb) [reply pkg/maps/cells.go:13]: Ah nvm utime.Cell is not a map.

Comment (mhofstetter) [reply pkg/maps/cells.go:13]: exactly - utime just syncs the utime and uses the configmap to write from user- to kernelspace. 

the cell of the configmap itself is part of the maps cell.

**Comment by mhofstetter** [comment]
rebased to main

**Comment by mhofstetter** [comment]
/test-1.27-net-next

**Comment by mhofstetter** [comment]
`test-1.25-5.4` (https://jenkins.cilium.io/job/Cilium-PR-K8s-1.25-kernel-5.4/107/) flaked with https://github.com/cilium/cilium/issues/24966

**Comment by michi-covalent** [comment]
why is this ready to merge? it's not approved by cilium/sig-datapath yet right?

**Comment by mhofstetter** [comment]
> why is this ready to merge? it's not approved by cilium/sig-datapath yet right?

@michi-covalent jussi is part of sig-datapath and looks like the perfect reviewer for this :)

**Comment by michi-covalent** [comment]
> @michi-covalent jussi is part of sig-datapath and looks like the perfect reviewer for this :)

oh yeah he is. sorry i got confused because yutaro is still listed as the code owner from sig-dapath.

**Pull request merged**
